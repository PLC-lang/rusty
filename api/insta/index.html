<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="What are snapshot tests"><title>insta - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-ba5701c5741a7b69.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="insta" data-themes="" data-resource-suffix="" data-rustdoc-version="1.70.0 (90c541806 2023-05-31)" data-search-js="search-e077946657036a58.js" data-settings-js="settings-298e1ea74db45b39.js" data-settings-css="settings-7bfb4c59cc6bc502.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../static.files/storage-62ce34ea385b278a.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-f61008743c98d196.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../insta/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../insta/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate insta</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 1.34.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">insta</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/insta/lib.rs.html#1-326">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><div align="center">
 <img src="https://github.com/mitsuhiko/insta/blob/master/assets/logo.png?raw=true" width="250" height="250">
 <p><strong>insta: a snapshot testing library for Rust</strong></p>
</div>
<h2 id="what-are-snapshot-tests"><a href="#what-are-snapshot-tests">What are snapshot tests</a></h2>
<p>Snapshots tests (also sometimes called approval tests) are tests that
assert values against a reference value (the snapshot).  This is similar
to how <code>assert_eq!</code> lets you compare a value against a reference value but
unlike simple string assertions, snapshot tests let you test against complex
values and come with comprehensive tools to review changes.</p>
<p>Snapshot tests are particularly useful if your reference values are very
large or change often.</p>
<h2 id="what-it-looks-like"><a href="#what-it-looks-like">What it looks like:</a></h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="attr">#[test]
</span><span class="kw">fn </span>test_hello_world() {
    <span class="macro">insta::assert_debug_snapshot!</span>(<span class="macro">vec!</span>[<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>]);
}</code></pre></div>
<p>Where are the snapshots stored?  Right next to your test in a folder
called <code>snapshots</code> as individual <a href="https://insta.rs/docs/snapshot-files/"><code>.snap</code> files</a>.</p>
<p>Got curious?</p>
<ul>
<li><a href="https://insta.rs/docs/quickstart/">Read the introduction</a></li>
<li><a href="https://insta.rs/docs/">Read the main documentation</a> which does not just
cover the API of the crate but also many of the details of how it works.</li>
<li>There is a screencast that shows the entire workflow: <a href="https://www.youtube.com/watch?v=rCHrMqE4JOY&amp;feature=youtu.be">watch the insta
introduction screencast</a>.</li>
</ul>
<h2 id="writing-tests"><a href="#writing-tests">Writing Tests</a></h2>
<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>insta::assert_debug_snapshot;

<span class="attr">#[test]
</span><span class="kw">fn </span>test_snapshots() {
    <span class="macro">assert_debug_snapshot!</span>(<span class="macro">vec!</span>[<span class="number">1</span>, <span class="number">2</span>, <span class="number">3</span>]);
}</code></pre></div>
<p>The recommended flow is to run the tests once, have them fail and check
if the result is okay.  By default the new snapshots are stored next
to the old ones with the extra <code>.new</code> extension.  Once you are satisfied
move the new files over.  To simplify this workflow you can use
<code>cargo insta review</code> (requires
<a href="https://crates.io/crates/cargo-insta"><code>cargo-insta</code></a>) which will let you
interactively review them:</p>
<div class="example-wrap"><pre class="language-text"><code>$ cargo test
$ cargo insta review
</code></pre></div><h2 id="assertion-macros"><a href="#assertion-macros">Assertion Macros</a></h2>
<p>This crate exports multiple macros for snapshot testing:</p>
<ul>
<li><a href="macro.assert_snapshot.html" title="macro insta::assert_snapshot"><code>assert_snapshot!</code></a> for comparing basic string snapshots.</li>
<li><a href="macro.assert_debug_snapshot.html" title="macro insta::assert_debug_snapshot"><code>assert_debug_snapshot!</code></a> for comparing <a href="https://doc.rust-lang.org/1.70.0/core/fmt/macros/derive.Debug.html" title="derive core::fmt::macros::Debug"><code>Debug</code></a> outputs of values.</li>
<li><a href="macro.assert_display_snapshot.html" title="macro insta::assert_display_snapshot"><code>assert_display_snapshot!</code></a> for comparing <a href="https://doc.rust-lang.org/1.70.0/core/fmt/trait.Display.html" title="trait core::fmt::Display"><code>Display</code></a> outputs of values.</li>
</ul>
<p>The following macros require the use of serde’s <a href="serde::Serialize"><code>Serialize</code></a>:</p>
<p>For macros that work with [<code>serde</code>] this crate also permits redacting of
partial values.  See <a href="https://insta.rs/docs/redactions/">redactions in the
documentation</a> for more information.</p>
<h2 id="snapshot-updating"><a href="#snapshot-updating">Snapshot updating</a></h2>
<p>During test runs snapshots will be updated according to the <code>INSTA_UPDATE</code>
environment variable.  The default is <code>auto</code> which will write all new
snapshots into <code>.snap.new</code> files if no CI is detected so that
<a href="https://crates.io/crates/cargo-insta"><code>cargo-insta</code></a>
can pick them up.  Normally you don’t have to change this variable.</p>
<p><code>INSTA_UPDATE</code> modes:</p>
<ul>
<li><code>auto</code>: the default. <code>no</code> for CI environments or <code>new</code> otherwise</li>
<li><code>always</code>: overwrites old snapshot files with new ones unasked</li>
<li><code>unseen</code>: behaves like <code>always</code> for new snapshots and <code>new</code> for others</li>
<li><code>new</code>: write new snapshots into <code>.snap.new</code> files</li>
<li><code>no</code>: does not update snapshot files at all (just runs tests)</li>
</ul>
<p>When <code>new</code> or <code>auto</code> is used as mode the <a href="https://crates.io/crates/cargo-insta"><code>cargo-insta</code></a>
command can be used to review the snapshots conveniently:</p>
<div class="example-wrap"><pre class="language-text"><code>$ cargo insta review
</code></pre></div>
<p>“enter” or “a” accepts a new snapshot, “escape” or “r” rejects,
“space” or “s” skips the snapshot for now.</p>
<p>For more information <a href="https://insta.rs/docs/cli/">read the cargo insta docs</a>.</p>
<h2 id="inline-snapshots"><a href="#inline-snapshots">Inline Snapshots</a></h2>
<p>Additionally snapshots can also be stored inline.  In that case the format
for the snapshot macros is <code>assert_snapshot!(reference_value, @&quot;snapshot&quot;)</code>.
The leading at sign (<code>@</code>) indicates that the following string is the
reference value.  <code>cargo-insta</code> will then update that string with the new
value on review.</p>
<p>Example:</p>

<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="attr">#[derive(Serialize)]
</span><span class="kw">pub struct </span>User {
   username: String,
}

<span class="macro">assert_yaml_snapshot!</span>(User {
   username: <span class="string">&quot;john_doe&quot;</span>.to_string(),
}, @<span class="string">&quot;&quot;</span>);</code></pre></div>
<p>Like with normal snapshots after the initial test failure you can run
<code>cargo insta review</code> to accept the change.  The file will then be updated
automatically.</p>
<h2 id="features"><a href="#features">Features</a></h2>
<p>The following features exist:</p>
<ul>
<li><code>csv</code>: enables CSV support (via serde)</li>
<li><code>json</code>: enables JSON support (via serde)</li>
<li><code>ron</code>: enables RON support (via serde)</li>
<li><code>toml</code>: enables TOML support (via serde)</li>
<li><code>yaml</code>: enables YAML support (via serde)</li>
<li><code>redactions</code>: enables support for redactions</li>
<li><code>filters</code>: enables support for filters</li>
<li><code>glob</code>: enables support for globbing ([<code>glob!</code>])</li>
<li><code>colors</code>: enables color output (enabled by default)</li>
</ul>
<p>For legacy reasons the <code>json</code> and <code>yaml</code> features are enabled by default
in limited capacity.  You will receive a deprecation warning if you are
not opting into them but for now the macros will continue to function.</p>
<p>Enabling any of the serde based formats enables the hidden <code>serde</code> feature
which gates some serde specific APIs such as [<code>Settings::set_info</code>].</p>
<h2 id="dependencies"><a href="#dependencies">Dependencies</a></h2>
<p><code>insta</code> tries to be light in dependencies but this is tricky to accomplish
given what it tries to do.  By default it currently depends on <code>serde</code> for
the [<code>assert_toml_snapshot!</code>] and [<code>assert_yaml_snapshot!</code>] macros.  In
the future this default dependencies will be removed.  To already benefit
from this optimization you can disable the default features and manually
opt into what you want.</p>
<h2 id="settings-1"><a href="#settings-1">Settings</a></h2>
<p>There are some settings that can be changed on a per-thread (and thus
per-test) basis.  For more information see <a href="struct.Settings.html" title="struct insta::Settings">Settings</a>.</p>
<p>Additionally Insta will load a YAML config file with settings that change
the behavior of insta between runs.  It’s loaded from any of the following
locations: <code>.config/insta.yaml</code>, <code>insta.yaml</code> and <code>.insta.yaml</code> from the
workspace root.  The following config options exist:</p>
<div class="example-wrap"><pre class="language-yaml"><code>behavior:
  # also set by INSTA_FORCE_UPDATE
  force_update: true/false
  # also set by INSTA_FORCE_PASS
  force_pass: true/false
  # also set by INSTA_OUTPUT
  output: &quot;diff&quot; | &quot;summary&quot; | &quot;minimal&quot; | &quot;none&quot;
  # also set by INSTA_UPDATE
  update: &quot;auto&quot; | &quot;always&quot; | &quot;new&quot; | &quot;unseen&quot; | &quot;no&quot;
  # also set by INSTA_GLOB_FAIL_FAST
  glob_fail_fast: true/false

# these are used by cargo insta test
test:
  # also set by INSTA_TEST_RUNNER
  runner: &quot;auto&quot; | &quot;cargo-test&quot; | &quot;nextest&quot;
  # automatically assume --review was passed to cargo insta test
  auto_review: true/false
  # automatically assume --accept-unseen was passed to cargo insta test
  auto_accept_unseen: true/false

# these are used by cargo insta review
review:
  # also look for snapshots in ignored folders
  include_ignored: true / false
  # also look for snapshots in hidden folders
  include_hidden: true / false
  # show a warning if undiscovered (ignored or hidden) snapshots are found.
  # defaults to true but creates a performance hit.
  warn_undiscovered: true / false
</code></pre></div><h2 id="optional-faster-runs"><a href="#optional-faster-runs">Optional: Faster Runs</a></h2>
<p>Insta benefits from being compiled in release mode, even as dev dependency.  It
will compile slightly slower once, but use less memory, have faster diffs and
just generally be more fun to use.  To achieve that, opt <code>insta</code> and <code>similar</code>
(the diffing library) into higher optimization in your <code>Cargo.toml</code>:</p>
<div class="example-wrap"><pre class="language-yaml"><code>[profile.dev.package.insta]
opt-level = 3

[profile.dev.package.similar]
opt-level = 3
</code></pre></div></div></details><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="internals/index.html" title="mod insta::internals">internals</a></div><div class="desc docblock-short">Exposes some library internals.</div></li></ul><h2 id="macros" class="small-section-header"><a href="#macros">Macros</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.allow_duplicates.html" title="macro insta::allow_duplicates">allow_duplicates</a></div><div class="desc docblock-short">Utility macro to permit a multi-snapshot run where all snapshots match.</div></li><li><div class="item-name"><a class="macro" href="macro.assert_debug_snapshot.html" title="macro insta::assert_debug_snapshot">assert_debug_snapshot</a></div><div class="desc docblock-short">Asserts a <code>Debug</code> snapshot.</div></li><li><div class="item-name"><a class="macro" href="macro.assert_display_snapshot.html" title="macro insta::assert_display_snapshot">assert_display_snapshot</a></div><div class="desc docblock-short">Asserts a <code>Display</code> snapshot.</div></li><li><div class="item-name"><a class="macro" href="macro.assert_snapshot.html" title="macro insta::assert_snapshot">assert_snapshot</a></div><div class="desc docblock-short">Asserts a string snapshot.</div></li><li><div class="item-name"><a class="macro" href="macro.with_settings.html" title="macro insta::with_settings">with_settings</a></div><div class="desc docblock-short">Settings configuration macro.</div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.MetaData.html" title="struct insta::MetaData">MetaData</a></div><div class="desc docblock-short">Snapshot metadata information.</div></li><li><div class="item-name"><a class="struct" href="struct.Settings.html" title="struct insta::Settings">Settings</a></div><div class="desc docblock-short">Configures how insta operates at test time.</div></li><li><div class="item-name"><a class="struct" href="struct.Snapshot.html" title="struct insta::Snapshot">Snapshot</a></div><div class="desc docblock-short">A helper to work with stored snapshots.</div></li></ul></section></div></main></body></html>