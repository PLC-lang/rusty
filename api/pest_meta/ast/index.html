<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Types for the pest’s abstract syntax tree."><title>pest_meta::ast - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../static.files/rustdoc-5bc39a1768837dd0.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="pest_meta" data-themes="" data-resource-suffix="" data-rustdoc-version="1.77.0 (aedd173a2 2024-03-17)" data-channel="1.77.0" data-search-js="search-dd67cee4cfa65049.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-48f368f3872407c8.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="icon" href="https://raw.githubusercontent.com/pest-parser/pest/master/pest-logo.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../pest_meta/index.html"><img src="https://raw.githubusercontent.com/pest-parser/pest/master/pest-logo.svg" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../pest_meta/index.html"><img src="https://raw.githubusercontent.com/pest-parser/pest/master/pest-logo.svg" alt="logo"></a><h2><a href="../../pest_meta/index.html">pest_meta</a><span class="version">2.7.9</span></h2></div><h2 class="location"><a href="#">Module ast</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li></ul></section><h2><a href="../index.html">In crate pest_meta</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../pest_meta/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../index.html">pest_meta</a>::<wbr><a class="mod" href="#">ast</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="../../src/pest_meta/ast.rs.html#10-694">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Types for the pest’s abstract syntax tree.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ExprTopDownIterator.html" title="struct pest_meta::ast::ExprTopDownIterator">ExprTopDownIterator</a></div><div class="desc docblock-short">The top down iterator for an expression.</div></li><li><div class="item-name"><a class="struct" href="struct.Rule.html" title="struct pest_meta::ast::Rule">Rule</a></div><div class="desc docblock-short">A grammar rule</div></li></ul><h2 id="enums" class="section-header">Enums<a href="#enums" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.Expr.html" title="enum pest_meta::ast::Expr">Expr</a></div><div class="desc docblock-short">All possible rule expressions</div></li><li><div class="item-name"><a class="enum" href="enum.RuleType.html" title="enum pest_meta::ast::RuleType">RuleType</a></div><div class="desc docblock-short">All possible rule types</div></li></ul></section></div></main></body></html>