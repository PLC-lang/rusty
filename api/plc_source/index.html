<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `plc_source` crate."><title>plc_source - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceSerif4-Bold-a2c9cd1067f8b328.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2"><link rel="stylesheet" href="../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../static.files/rustdoc-ba5701c5741a7b69.css" id="mainThemeStyle"><div id="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="plc_source" data-themes="" data-resource-suffix="" data-rustdoc-version="1.70.0 (90c541806 2023-05-31)" data-search-js="search-e077946657036a58.js" data-settings-js="settings-298e1ea74db45b39.js" data-settings-css="settings-7bfb4c59cc6bc502.css" data-theme-light-css="light-0f8c037637f9eb3e.css" data-theme-dark-css="dark-1097f8e92a01e3cf.css" data-theme-ayu-css="ayu-614652228113ac93.css" ></div><script src="../static.files/storage-62ce34ea385b278a.js"></script><script defer src="../crates.js"></script><script defer src="../static.files/main-f61008743c98d196.js"></script><noscript><link rel="stylesheet" media="(prefers-color-scheme:light)" href="../static.files/light-0f8c037637f9eb3e.css"><link rel="stylesheet" media="(prefers-color-scheme:dark)" href="../static.files/dark-1097f8e92a01e3cf.css"><link rel="stylesheet" href="../static.files/noscript-13285aec31fa243e.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod crate"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="logo-container" href="../plc_source/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2></h2></nav><nav class="sidebar"><a class="logo-container" href="../plc_source/index.html"><img class="rust-logo" src="../static.files/rust-logo-151179464ae7ed46.svg" alt="logo"></a><h2 class="location"><a href="#">Crate plc_source</a></h2><div class="sidebar-elems"><ul class="block"><li class="version">Version 0.1.0</li><li><a id="all-types" href="all.html">All Items</a></li></ul><section><ul class="block"><li><a href="#modules">Modules</a></li><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Definitions</a></li></ul></section></div></nav><main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><a href="../help.html">?</a></div><div id="settings-menu" tabindex="-1"><a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Crate <a class="mod" href="#">plc_source</a><button id="copy-path" title="Copy item path to clipboard"><img src="../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="srclink" href="../src/plc_source/lib.rs.html#1-214">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><h2 id="modules" class="small-section-header"><a href="#modules">Modules</a></h2><ul class="item-table"><li><div class="item-name"><a class="mod" href="source_location/index.html" title="mod plc_source::source_location">source_location</a></div></li></ul><h2 id="structs" class="small-section-header"><a href="#structs">Structs</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.SourceCode.html" title="struct plc_source::SourceCode">SourceCode</a></div><div class="desc docblock-short">The SourceCode unit is the smallest unit of compilation that can be passed to the compiler</div></li></ul><h2 id="enums" class="small-section-header"><a href="#enums">Enums</a></h2><ul class="item-table"><li><div class="item-name"><a class="enum" href="enum.SourceType.html" title="enum plc_source::SourceType">SourceType</a></div><div class="desc docblock-short">Represents the type of source a SourceContainer holds</div></li></ul><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.Compilable.html" title="trait plc_source::Compilable">Compilable</a></div></li><li><div class="item-name"><a class="trait" href="trait.SourceCodeFactory.html" title="trait plc_source::SourceCodeFactory">SourceCodeFactory</a></div><div class="desc docblock-short">Extension trait to create sources with names from strs, used in tests</div></li><li><div class="item-name"><a class="trait" href="trait.SourceContainer.html" title="trait plc_source::SourceContainer">SourceContainer</a></div><div class="desc docblock-short">SourceContainers offer source-code to be compiled via the load_source function.
Furthermore it offers a location-String used when reporting diagnostics.</div></li></ul><h2 id="functions" class="small-section-header"><a href="#functions">Functions</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.create_source_code.html" title="fn plc_source::create_source_code">create_source_code</a></div></li></ul><h2 id="types" class="small-section-header"><a href="#types">Type Definitions</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.BuildDescriptionSource.html" title="type plc_source::BuildDescriptionSource">BuildDescriptionSource</a></div></li></ul></section></div></main></body></html>