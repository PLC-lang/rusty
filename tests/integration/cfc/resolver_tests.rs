use driver::parse_and_annotate;
use insta::assert_debug_snapshot;
use plc_ast::ast::{Assignment, AstStatement};
use plc_source::SourceContainer;
use rusty::resolver::AnnotationMap;

use crate::get_test_file;

#[test]
fn label_added_to_index_as_annotation() {
    let cfc_file = get_test_file("cfc/jump_true.cfc");
    let mut cfc_file = cfc_file.load_source(None).unwrap();
    //Remove the path
    cfc_file.path.replace("<internal>.cfc".into());

    let index = parse_and_annotate("plc", vec![cfc_file]).unwrap().index;
    assert_debug_snapshot!(index.get_label("main", "lbl").unwrap());
}

#[test]
fn jumps_annotated_with_label_annoations() {
    let cfc_file = get_test_file("cfc/jump_true.cfc");
    let mut cfc_file = cfc_file.load_source(None).unwrap();
    //Remove the path
    cfc_file.path.replace("<internal>.cfc".into());

    let annotated_project = parse_and_annotate("plc", vec![cfc_file]).unwrap();
    let annotations = &annotated_project.annotations;
    let (unit, ..) = &annotated_project.units[0];
    // Get the jump
    let jump = &unit.implementations[0].statements[1];
    assert_debug_snapshot!(annotations.get(jump))
}

#[test]
fn unbound_jumps_not_annotated() {
    let cfc_file = get_test_file("cfc/jump_no_label.cfc");
    let mut cfc_file = cfc_file.load_source(None).unwrap();
    //Remove the path
    cfc_file.path.replace("<internal>.cfc".into());

    let annotated_project = parse_and_annotate("plc", vec![cfc_file]).unwrap();
    let annotations = &annotated_project.annotations;
    let (unit, ..) = &annotated_project.units[0];
    // Get the jump
    let jump = &unit.implementations[0].statements[1];
    assert!(annotations.get(jump).is_none())
}

#[test]
fn action_variables_annotated() {
    let cfc_file = get_test_file("cfc/actions.cfc");
    let mut cfc_file = cfc_file.load_source(None).unwrap();
    //Remove the path
    cfc_file.path.replace("<internal>.cfc".into());

    let annotated_project = parse_and_annotate("plc", vec![cfc_file]).unwrap();
    let annotations = &annotated_project.annotations;
    let (unit, ..) = &annotated_project.units[0];

    //Action 1 and 2 calls annotated
    let act1 = &unit.implementations[0].statements[1];
    assert_debug_snapshot!(annotations.get(act1));
    let act2 = &unit.implementations[0].statements[2];
    assert_debug_snapshot!(annotations.get(act2));
    //In action 1 a is annotated
    let AstStatement::Assignment(Assignment { left, .. }) = &unit.implementations[1].statements[0].get_stmt()
    else {
        unreachable!("Statement must be an assingment");
    };
    assert_debug_snapshot!(annotations.get(left));
    //In action 2 b is annotated
    let AstStatement::Assignment(Assignment { left, .. }) = &unit.implementations[2].statements[0].get_stmt()
    else {
        unreachable!("Statement must be an assingment");
    };
    assert_debug_snapshot!(annotations.get(left));
}
