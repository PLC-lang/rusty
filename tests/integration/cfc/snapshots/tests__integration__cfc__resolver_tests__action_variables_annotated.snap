---
source: tests/integration/cfc/resolver_tests.rs
expression: annotations.get(act1)
---
Some(
    Program {
        qualified_name: "main.newAction",
    },
)
